### PR TITLE
lib/nolibc: Fix memory bug in vfscanf

### PR DESCRIPTION
**Pull request by thass0**
It wasn't guaranteed that the `buf` buffer would be null-terminated in all cases. The fix makes sure that there is always as null byte at the end.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]

### CODE REVIEW

**Review by michpappas — CHANGES_REQUESTED** [review]
Good catch 💯 Perhaps it would be reasonable for the termination to be performed by the function that reads the input.

**Comment by thass0** [comment]
> Good catch 💯 Perhaps it would be reasonable for the termination to be performed by the function that reads the input.

I agree. I'm moving to to `uk_scanf`. I added an assertion in `uk_scanf`. The condition that's asserted should always hold given that `count` is initially assigned `*cnt` and not touched after that and the `while`-loop makes sure `bytes_total < count`. But better to assert this before writing out of bounds of the buffer.

**Comment by thass0** [comment]
You're right, I moved the NULL-termination into `uk_scanf` itself. The assertion that I added should always hold given how `uk_scanf` is coded right now. I just added it to make sure that if something changes in the implementation of `uk_scanf`, we don't accidentally write outside the buffer.

**Comment by michpappas** [comment]
> You're right, I moved the NULL-termination into `uk_scanf` itself. The assertion that I added should always hold given how `uk_scanf` is coded right now. I just added it to make sure that if something changes in the implementation of `uk_scanf`, we don't accidentally write outside the buffer.

Regarding the assertion, I would feel better if we had confidence that the implementation is correct. Is that the case? Moreover, with the proposed change we overwrite the last character passed by the user, if the user writes the maximum number of characters. Perhaps it would be better if we reserved the last byte in the first place, and allow the user to send `cnt - 1` characters?

**Comment by thass0** [comment]
I pushed another update. Now we reserve a byte at the end. I also removed the assertion. The logic is this: the `cnt` parameter contains the number of bytes available in the buffer. When the `do { ... } while` loop finishes, `bytes_total` is at most equal to `cnt - 1`. So it's safe to use `bytes_total` to index into `buffer` (which is `cnt` bytes in size). `bytes_total` can't be bigger than `cnt - 1` when the loops finishes because each iteration of the loop increases `bytes_total` by at most `(cnt - 1) - bytes_total`.

File: lib/nolibc/scanf.c
Comment (michpappas) [inline lib/nolibc/scanf.c]: Now we removed the assertion we probably don't need this. Other than that it LGTM 👍🏼

Comment (thass0) [reply lib/nolibc/scanf.c]: Right, the include is gone now

**Comment by michpappas** [comment]
Actually that function still doesn't look really robust. Consider for example what happens when the caller passes `cnt = 1`. The same thing happens if the caller input the same amount of characters as `sizeof(buffer) - 1`.

**Comment by thass0** [comment]
If `cnt` is passed as one, then we run into an endless loop, right? Because `count` is `0` then so `ukplat_cink` will also return `0`. But the loop around `ukplat_cink` only ends if `ukplat_cink` returns a value greater than zero.

> The same thing happens if the caller input the same amount of characters as `sizeof(buffer) - 1`.

What happens exactly?

**Comment by michpappas** [comment]
> If cnt is passed as one, then we run into an endless loop, right? 

Exactly, it runs into an endless loop. So your fix needs a bit more work to address that problem. With the opportunity you can also update that loop to handle `EOF` (Ctrl-D).

**Review by michpappas — CHANGES_REQUESTED** [review]
Thanks. Overall this function could take a bit of rework, and we would probably need a tty layer to handle things like EOF, but until then let's integrate this fix to make sure that at least the buffer is terminated.

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

**Comment by razvand** [comment]
@andreittr , could you also review this PR?

File: lib/nolibc/scanf.c
Comment (skuenzer) [inline lib/nolibc/scanf.c:20]: Does this comment fit with the check below (`at least two bytes` != `<= 1`)?
In the case of `size = 0`, couldn't we return without reading a char from ukplat_cink?

**Comment by skuenzer** [comment]
> > If cnt is passed as one, then we run into an endless loop, right?
> 
> Exactly, it runs into an endless loop. So your fix needs a bit more work to address that problem. With the opportunity you can also update that loop to handle `EOF` (Ctrl-D).

Saw this, too. This is fine, we can address later.  Probably we could in the case of `size = 1`, just return without reading a char from ukplat_cink (since there is no space).

**Comment by skuenzer** [comment]
> @andreittr , could you also review this PR?

I think it this small bugfix PR is fine without a second reviewer.

**Review by skuenzer — APPROVED** [review]
Thanks!

Approved-by: Simon Kuenzer <simon@unikraft.io>

**Comment by skuenzer** [comment]
> Thanks. Overall this function could take a bit of rework, and we would probably need a tty layer to handle things like EOF, but until then let's integrate this fix to make sure that at least the buffer is terminated.
> 
> Reviewed-by: Michalis Pappas [michalis@unikraft.io](mailto:michalis@unikraft.io)

@michpappas Can you send this tag with a GitHub approval? ...otherwise our tooling will block merging the PR.

**Review by michpappas — APPROVED** [review]
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

**Pull request merged**
